### PR TITLE
feat(CheckedFile): now support a char* input

### PR DESCRIPTION
**Pull request by jean-noelp**
(#15)

Now you can directly use ImageFile like this:

`char *buffer;`
`[...] // read file`
`ImageFile imf(buffer, size);`

### CODE REVIEW

File: src/CheckedFile.cpp [hunk 1]
```diff
@@ -384,25 +437,38 @@ void CheckedFile::seek(uint64_t offset, OffsetMode omode)
 
 uint64_t CheckedFile::lseek64(int64_t offset, int whence)
 {
+   int64_t result = -1;
+   if(fd_ >= 0){
 #if defined(_WIN32)
 #  if defined(_MSC_VER) || defined(__MINGW32__) //<rs 2010-06-16> mingw _is_ WIN32!
-   __int64 result = _lseeki64(fd_, offset, whence);
```
Comment (asmaloney) [inline src/CheckedFile.cpp]: You're changing the return type for this. I would prefer not to. `_lseeki64` returns `__int64`.

If you want to unify these in another PR, maybe there's a better function to use than `_lseeki64` on Windows?

File: src/CheckedFile.cpp [hunk 2]
```diff
@@ -384,25 +437,38 @@ void CheckedFile::seek(uint64_t offset, OffsetMode omode)
...
-   int64_t result = ::lseek64(fd_, offset, whence);
+   result = ::lseek64(fd_, offset, whence);
 #elif defined(__APPLE__)
-   int64_t result = ::lseek(fd_, offset, whence);
+   result = ::lseek(fd_, offset, whence);
 #else
 #  error "no supported OS platform defined"
 #endif
+   }
+   else if(bufView_ != nullptr){
```
Comment (asmaloney) [inline src/CheckedFile.cpp]: Style: space after keywords
Style: `{` on its own line

I won't point these out everywhere, but please change throughout the PR...

File: src/CheckedFile.cpp [hunk 3]
```diff
@@ -384,25 +437,38 @@ void CheckedFile::seek(uint64_t offset, OffsetMode omode)
...
+      if(bufView_->seek(offset, whence)){
+         result = bufView_->pos();
+      }
+      else{
+         cerr << "CheckedFile::lseek64 Error: cursorStream_ is out of bound!" << endl;
```
Comment (asmaloney) [inline src/CheckedFile.cpp]: Please remove.

This already throws an exception (next block). None of the code in this file is outputting any errors on `std::cerr`.

File: src/CheckedFile.cpp [hunk 4]
```diff
@@ -565,6 +631,14 @@ void CheckedFile::close()
 
       fd_ = -1;
    }
+
+   if (bufView_ != nullptr) {
+      delete bufView_;
+      bufView_ = nullptr;
+
+      // WARNING: do NOT delete buffer of streamView_ because
```
Comment (asmaloney) [inline src/CheckedFile.cpp]: `streamView_`? Do you mean `bufView_`?

If so, this comment should be in the BufferView class, not here. Maybe just add a comment on the BufferView constructor that the caller retains ownership of the pointer.

File: src/CheckedFile.cpp [hunk 5]
```diff
@@ -658,13 +732,20 @@ void CheckedFile::readPhysicalPage(char* page_buffer, uint64_t page)
    /// Seek to start of physical page
    seek( page*physicalPageSize, Physical );
 
+   size_t result = -1;
+   if(fd_ >= 0){
 #if defined(_MSC_VER)
-   int result = ::_read( fd_, page_buffer, physicalPageSize );
```
Comment (asmaloney) [inline src/CheckedFile.cpp]: Same comment as above - please don't change the return types.

`size_t` is not the same as an `int`.

File: src/CheckedFile.h [hunk 6]
```diff
@@ -32,6 +32,20 @@
 #include "Common.h"
 
 namespace e57 {
+   /// Tool class to read buffer efficiently without 
+   /// multiplying copy operations
+   class BufferView {
```
Comment (asmaloney) [inline src/CheckedFile.h]: This class declaration should be in `CheckedFile.cpp` and forward declared here.

**Comment by jean-noelp** [comment]
I pushed an other commit to answer to all your requests ^^

File: src/CheckedFile.cpp [hunk 7]
```diff
@@ -80,6 +80,69 @@ constexpr size_t     CheckedFile::physicalPageSize;
...
+class e57::BufferView
+{
+   public:
+      /// @param[IN] input: filled buffer owned by caller.
+      /// @param[IN] size: size of input
+      BufferView( const char* input, uint64_t size ) : 
+         stream_( input ),
+         streamSize_( size )
+      {
+
```
Comment (asmaloney) [inline src/CheckedFile.cpp]: Style: remove blank line

File: src/CheckedFile.cpp [hunk 8]
```diff
@@ -565,6 +668,14 @@ void CheckedFile::close()
 
       fd_ = -1;
    }
+
+   if (bufView_ != nullptr) {
```
Comment (asmaloney) [inline src/CheckedFile.cpp]: Style: `{` on new line

File: src/CheckedFile.cpp [hunk 9]
```diff
@@ -658,10 +769,23 @@ void CheckedFile::readPhysicalPage(char* page_buffer, uint64_t page)
...
+      {
+         throw E57_EXCEPTION2(E57_ERROR_READ_FAILED, "fileName=" + fileName_ + " result=" + toString(result));
+      }
+
+      return;
+   }
+
 #if defined(_MSC_VER)
-   int result = ::_read( fd_, page_buffer, physicalPageSize );
+      __int64 result = result = ::_read( fd_, page_buffer, physicalPageSize );
```
Comment (asmaloney) [inline src/CheckedFile.cpp]: Looks like a merge mistake? Should simply be `int result = ...`

File: src/CheckedFile.cpp [hunk 10]
```diff
@@ -658,10 +769,23 @@ void CheckedFile::readPhysicalPage(char* page_buffer, uint64_t page)
...
 #elif defined(__GNUC__)
-   ssize_t result = ::read( fd_, page_buffer, physicalPageSize );
+      ssize_t result = ::read( fd_, page_buffer, physicalPageSize );
```
Comment (asmaloney) [inline src/CheckedFile.cpp]: Style: Why the extra indent?

**Comment by jean-noelp** [comment]
3rd time will be the good one! :D

**Comment by asmaloney** [comment]
Thank you for your contribution!

**Comment by jean-noelp** [comment]
You're welcome!
Your reactivity deserved a PR, at least!

**Pull request merged**